tools/clang_tools/api_booster/proto_cxx_utils.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include <regex>
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+namespace ApiBooster {
+
+// Protobuf C++ code generation hackery. This is where the utilities that map
+// between C++ and protobuf types, enum constants and identifiers live. Most of
+// this is heuristic and needs to match whatever the protobuf compiler does.
+class ProtoCxxUtils {
+public:
+  // Convert from a C++ type, e.g. foo::bar::v2, to a protobuf type, e.g.
+  // foo.bar.v2.
+  static std::string cxxToProtoType(const std::string& cxx_type_name) {
+    // Convert from C++ to a qualified proto type. This is fairly hacky stuff,
```

Comment:
Is it possible to read this from proto descriptor that embedded in `pb.cc` source code? I guess it is not that easy in this context?

Reply:
Not sure how that would work, since there is still not a direct associative map between the proto and C++ types. I think we could be more robust if we did consult the proto descriptor in addition, but that's probably P2 IMHO in terms of importance at this stage.

Reply:
Also, it's worth noting that types aren't the only thing that need munging here; all the C++ generated methods and statics, e.g. `Foo_Parse()` or `.mutable_foo()` need similar kinds of reverse engineering. I think the conventions permit this all using string operations, but AFAIK there is no reflection possible on this.